src/Build/BackEnd/Node/RarNode.cs

```diff
@@ -0,0 +1,93 @@
+using System.Threading.Tasks;
+
+using Microsoft.Build.BackEnd;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Internal;
+using Microsoft.Build.Shared;
+using Microsoft.Build.Tasks.ResolveAssemblyReferences.Server;
+
+namespace Microsoft.Build.Execution
+{
+    public sealed class RarNode : INode
+    {
+        /// <summary>
+        /// The amount of time to wait for the client to connect to the host.
+        /// </summary>
+        private const int ClientConnectTimeout = 60000;
+
+        public NodeEngineShutdownReason Run(bool nodeReuse, bool lowPriority, out Exception shutdownException)
+        {
+            shutdownException = null;
+            using CancellationTokenSource cts = new CancellationTokenSource();
+            string pipeName = CommunicationsUtilities.GetRarPipeName(nodeReuse, lowPriority);
+            RarController controller = new RarController(pipeName, NamedPipeUtil.CreateNamedPipeServer);
+
+            Task<int> rarTask = controller.StartAsync(cts.Token);
+
+            Handshake handshake = NodeProviderOutOfProc.GetHandshake(nodeReuse, enableLowPriority: lowPriority, specialNode: true);
+            Task<NodeEngineShutdownReason> msBuildShutdown = RunShutdownCheckAsync(handshake, cts.Token);
+
+            int index = Task.WaitAny(msBuildShutdown, rarTask);
+            cts.Cancel();
+
+            if (index == 0)
+            {
+                // We know that this task is completed so we can get Result without worring about waiting for it
+                return msBuildShutdown.Result;
+            }
+            else
+            {
+                // RAR task can only exit with connection error or timeout
+                return NodeEngineShutdownReason.ConnectionFailed;
+            }
+        }
+
+        public NodeEngineShutdownReason Run(out Exception shutdownException)
+        {
+            return Run(false, false, out shutdownException);
+        }
+
+        private async Task<NodeEngineShutdownReason> RunShutdownCheckAsync(Handshake handshake, CancellationToken cancellationToken = default)
+        {
+            string pipeName = NamedPipeUtil.GetPipeNameOrPath("MSBuild" + Process.GetCurrentProcess().Id);
+
+            while (true)
+            {
+                if (cancellationToken.IsCancellationRequested)
+                    return NodeEngineShutdownReason.Error;
+
+                using NamedPipeServerStream serverStream = NamedPipeUtil.CreateNamedPipeServer(pipeName, maxNumberOfServerInstances: NamedPipeServerStream.MaxAllowedServerInstances);
+                await serverStream.WaitForConnectionAsync(cancellationToken).ConfigureAwait(false);
+
+                bool connected = NamedPipeUtil.ValidateHandshake(handshake, serverStream, ClientConnectTimeout);
+                if (!connected)
+                    continue;
+
+                byte[] header = new byte[5];
+                int bytesRead = await serverStream.ReadAsync(header, 0, header.Length).ConfigureAwait(false);
```

Comment:
Can you explain this part to me?

Reply:
Packets have headers which are 5 bytes long. So we read them here and from first byte we found out what type of packet it is.

Reply:
Would you mind adding a comment? That makes sense in retrospect, but that wasn't obvious to me before.

Reply:
Added.